File: Python/Reorder_List.py
# LeetCode link: https://leetcode.com/problems/reorder-list/
# Issue: https://github.com/noisefilter19/LeetCode_Algorithms/issues/522
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: ListNode) -> None:
        curr = head
        temp_array = []
        while curr:
            temp_array.append(curr)
            curr = curr.next
        left = 0
        right = len(temp_array) - 1
        curr = head
        for _ in range(len(temp_array)):
            if _ % 2 == 0:
                curr.next = temp_array[left]
                left += 1
            else:
                curr.next = temp_array[right]
                right -= 1
            curr = curr.next
            if left > right:
                curr.next = None

File: C++/Reorder_list.cpp
class Solution {
public:
    void reorderList(ListNode* head) {
        if(head==NULL)
            return;
        ListNode* slow=head,*fast=head->next;
        while(fast && fast->next)
        {
            slow=slow->next;
            fast=fast->next->next;
        }
        ListNode* rev=slow->next;
        slow->next=NULL;
        ListNode* curr=rev,*prev=NULL,*next;
        while(curr!=NULL)
        {
            next=curr->next;
            curr->next=prev;
            prev=curr;
            curr=next;
        }
        rev=head;
        fast=head->next;
        int flag=0;
        while(fast || prev)
        {
            if(flag==0)
            {
                rev->next=prev;
                prev=prev->next;
                rev=rev->next;
                flag=1;
            }
            else
            {
                rev->next=fast;
                fast=fast->next;
                rev=rev->next;
                flag=0;
            }
        }
    }
};
